File: Code/BeforeLegends/Assets/Scripts/World Map/Controls/MouseTileInput.js
//#pragma strict

//static var instance : MouseTileInput;

//var lastTile : Vec2i = Vec2i(0, 0); 

//var lastTilePos : Vector3 = Vector3.zero;

//var lastWorldPos : Vector3 = Vector3.zero;

//function Awake(){
//	if(!instance) instance = this;
//}

//function Update(){
//	var hit : RaycastHit;
//	var data : WorldMapData = WorldMapData.getInstance();
//	if(Physics.Raycast(Camera.main.ScreenPointToRay(Input.mousePosition), hit)){
//		var newTileY : int = Mathf.Round(hit.point.z / (data.flatHex.size.z * 0.75));
//		if(newTileY >= 0 && newTileY < data.size.y){
//			var odd : boolean = newTileY % 2 == 1;
//			var newTileX : int = Mathf.Round((hit.point.x - (odd ? data.flatHex.extents.x : 0)) / data.flatHex.size.x);
//			if(newTileX >= 0 && newTileX < data.size.x){
//				lastWorldPos = hit.point;
//				if(newTileX != lastTile.x || newTileY != lastTile.y){
//					Messenger.instance.send(MouseTileChangedMessage(lastTile, lastTilePos, lastWorldPos));
//				}
//				lastTile.y = newTileY;
//				lastTile.x = newTileX;
//				lastTilePos = data.tiles[newTileX, newTileY].position;
//			}
//		}
//	}
//}	

File: Code/BeforeLegends/Assets/Scripts/World Map/Controls/MouseTileInput.cs
using UnityEngine;
using System.Collections;

public class MouseTileInput : MonoBehaviour {

    private static MouseTileInput _instance;

    public static MouseTileInput instance
    {
        get
        {
            if (_instance == null)
                _instance = GameObject.FindObjectOfType<MouseTileInput>();
            return _instance;
        }
    }

    public Vec2int lastTile = new Vec2int(0, 0); 

    public Vector3 lastTilePos = Vector3.zero;

    public Vector3  lastWorldPos = Vector3.zero;

    void Update(){
	    RaycastHit hit;
	    WorldMapData data = WorldMapData.instance;
        Ray r = Camera.main.ScreenPointToRay(Input.mousePosition);
	    if(Physics.Raycast(r, out hit))
        {
		    int newTileY = (int)Mathf.Round(hit.point.z / (data.flatHex.size.z * 0.75f));
		    if(newTileY >= 0 && newTileY < data.size.y){
			    bool odd = newTileY % 2 == 1;
                int newTileX = (int)Mathf.Round((hit.point.x - (odd ? data.flatHex.extents.x : 0)) / data.flatHex.size.x);
			    if(newTileX >= 0 && newTileX < data.size.x){
				    lastWorldPos = hit.point;
				    if(newTileX != lastTile.x || newTileY != lastTile.y){
					    Messenger.instance.send(new MouseTileChangedMessage(lastTile, lastTilePos, lastWorldPos));
				    }
				    lastTile.y = newTileY;
				    lastTile.x = newTileX;
				    lastTilePos = data.tiles[newTileX, newTileY].position;
			    }
		    }
	    }
    }	
}
